File: leetcode/11.盛最多水的容器.go
/*
 * @lc app=leetcode.cn id=11 lang=golang
 *
 * [11] 盛最多水的容器
 */

// @lc code=start
func maxArea(height []int) (cap int) {
	left, right := 0, len(height) - 1
	for left < right {
		area := min(height[left], height[right]) * (right-left)
		cap = max(cap, area)
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}
	return
}
func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end

File: leetcode/11.盛最多水的容器.rs
/*
 * @lc app=leetcode.cn id=11 lang=rust
 *
 * [11] 盛最多水的容器
 */

// @lc code=start
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut ret = 0;
        let (mut left, mut right) = (0i32, height.len() as i32 - 1);
        while left < right {
            let area = height[left as usize].min(height[right as usize]) * (right-left);
            ret = ret.max(area);
            if height[left as usize] < height[right as usize] {
                left += 1;
            } else {
                right -= 1;
            }
        }
        ret
    }
}
// @lc code=end
